File: 0066.js
/**
 * @param {number[]} digits
 * @return {number[]}
 */
var plusOne = function(digits) {
  let carry = 1;

  for (let i = digits.length - 1; i >= 0; i--) {
    if (carry === 0) {
      break;
    }

    if (digits[i] === 9) {
      digits[i] = 0;
    } else {
      digits[i] += 1;
      carry = 0;
    }
  }

  if (carry === 1) {
    digits.unshift(1);
  }

  return digits;
};

// case 1
let digits = [1, 2, 3];
let result = plusOne(digits);
console.log(result); // [1, 2, 4]

// case 2
digits = [4, 3, 2, 1];
result = plusOne(digits);
console.log(result); // [4, 3, 2, 2]

// case 3
digits = [9];
result = plusOne(digits);
console.log(result); // [1, 0]

// case 4
digits = [1, 9, 4];
result = plusOne(digits);
console.log(result); // [1, 9, 5]

File: 0066.py
from typing import List

class Solution:
  def plusOne(self, digits: List[int]) -> List[int]:
      digits = digits[::-1]
      one, i = 1, 0

      while one:
        if i < len(digits):
          if digits[i] == 9:
            digits[i] = 0
          else:
            digits[i] += 1
            one = 0
        else:
          digits.append(1)
          one = 0
        i += 1
      return digits[::-1]

solution = Solution()

# case 1
digits = [1, 2, 3]
result = solution.plusOne(digits)
print(result)

# case 2
digits = [4, 3, 2, 1]
result = solution.plusOne(digits)
print(result)

# case 3
digits = [9]
result = solution.plusOne(digits)
print(result)